File: concepts/08-hash-tables/node/hashTable.js
/**
 * Hash Table Implementation in Node.js
 */

class HashNode {
    /**
     * Node for chaining
     */
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

class HashTable {
    /**
     * Hash Table with Chaining
     */
    constructor(capacity = 10, loadFactor = 0.75) {
        this.capacity = capacity;
        this.loadFactor = loadFactor;
        this.size = 0;
        this.buckets = new Array(capacity).fill(null);
    }

    /**
     * Hash function - division method
     */
    _hash(key) {
        if (typeof key === 'number') {
            return key % this.capacity;
        }
        // For strings
        let hashVal = 0;
        for (const char of String(key)) {
            hashVal = (hashVal * 31 + char.charCodeAt(0)) % this.capacity;
        }
        return hashVal;
    }

    /**
     * Resize and rehash - O(n)
     */
    _resize() {
        const oldBuckets = this.buckets;
        this.capacity *= 2;
        this.buckets = new Array(this.capacity).fill(null);
        this.size = 0;

        for (const bucket of oldBuckets) {
            let current = bucket;
            while (current) {
                this.put(current.key, current.value);
                current = current.next;
            }
        }
    }

    /**
     * Insert/Update - O(1) average, O(n) worst
     */
    put(key, value) {
        const index = this._hash(key);

        // Check if key exists
        let current = this.buckets[index];
        while (current) {
            if (current.key === key) {
                current.value = value;
                return;
            }
            current = current.next;
        }

        // Insert new node
        const newNode = new HashNode(key, value);
        newNode.next = this.buckets[index];
        this.buckets[index] = newNode;
        this.size++;

        // Check load factor
        if (this.size > this.capacity * this.loadFactor) {
            this._resize();
        }
    }

    /**
     * Get value - O(1) average, O(n) worst
     */
    get(key) {
        const index = this._hash(key);
        let current = this.buckets[index];

        while (current) {
            if (current.key === key) {
                return current.value;
            }
            current = current.next;
        }

        return null;
    }

    /**
     * Remove key - O(1) average, O(n) worst
     */
    remove(key) {
        const index = this._hash(key);
        let current = this.buckets[index];
        let prev = null;

        while (current) {
            if (current.key === key) {
                if (prev) {
                    prev.next = current.next;
                } else {
                    this.buckets[index] = current.next;
                }
                this.size--;
                return true;
            }
            prev = current;
            current = current.next;
        }

        return false;
    }

    /**
     * Check if key exists - O(1) average
     */
    contains(key) {
        return this.get(key) !== null;
    }

    get length() {
        return this.size;
    }
}

class HashTableOpenAddressing {
    /**
     * Hash Table with Linear Probing
     */
    constructor(capacity = 10) {
        this.capacity = capacity;
        this.size = 0;
        this.keys = new Array(capacity).fill(null);
        this.values = new Array(capacity).fill(null);
        this.DELETED = Symbol('DELETED');
    }

    _hash(key) {
        if (typeof key === 'number') {
            return key % this.capacity;
        }
        let hashVal = 0;
        for (const char of String(key)) {
            hashVal = (hashVal * 31 + char.charCodeAt(0)) % this.capacity;
        }
        return hashVal;
    }

    _probe(key) {
        let index = this._hash(key);
        const initialIndex = index;

        while (this.keys[index] !== null &&
            this.keys[index] !== this.DELETED &&
            this.keys[index] !== key) {
            index = (index + 1) % this.capacity;
            if (index === initialIndex) {
                return -1;
            }
        }
        return index;
    }

    put(key, value) {
        if (this.size >= this.capacity) {
            return false;
        }

        const index = this._probe(key);
        if (index === -1) {
            return false;
        }

        if (this.keys[index] === null || this.keys[index] === this.DELETED) {
            this.size++;
        }

        this.keys[index] = key;
        this.values[index] = value;
        return true;
    }

    get(key) {
        let index = this._hash(key);
        const initialIndex = index;

        while (this.keys[index] !== null) {
            if (this.keys[index] === key) {
                return this.values[index];
            }
            index = (index + 1) % this.capacity;
            if (index === initialIndex) {
                break;
            }
        }

        return null;
    }

    remove(key) {
        let index = this._hash(key);
        const initialIndex = index;

        while (this.keys[index] !== null) {
            if (this.keys[index] === key) {
                this.keys[index] = this.DELETED;
                this.values[index] = null;
                this.size--;
                return true;
            }
            index = (index + 1) % this.capacity;
            if (index === initialIndex) {
                break;
            }
        }

        return false;
    }
}

// Example usage
if (require.main === module) {
    // Chaining
    console.log("--- Hash Table with Chaining ---");
    const ht = new HashTable();
    ht.put("apple", 1);
    ht.put("banana", 2);
    ht.put("cherry", 3);

    console.log("Get 'apple':", ht.get("apple"));
    console.log("Contains 'banana':", ht.contains("banana"));
    console.log("Size:", ht.length);

    ht.remove("banana");
    console.log("After removing 'banana', contains:", ht.contains("banana"));

    // Open Addressing
    console.log("\n--- Hash Table with Open Addressing ---");
    const ht2 = new HashTableOpenAddressing();
    ht2.put(1, "one");
    ht2.put(2, "two");
    ht2.put(11, "eleven");

    console.log("Get 1:", ht2.get(1));
    console.log("Get 11:", ht2.get(11));
    ht2.remove(1);
    console.log("After removing 1, get 1:", ht2.get(1));
}

module.exports = { HashTable, HashTableOpenAddressing };

File: concepts/08-hash-tables/python/hash_table.py
"""
Hash Table Implementation in Python
"""

class HashNode:
    """Node for chaining"""
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None


class HashTable:
    """Hash Table with Chaining"""
    
    def __init__(self, capacity=10, load_factor=0.75):
        self.capacity = capacity
        self.load_factor = load_factor
        self.size = 0
        self.buckets = [None] * capacity
    
    def _hash(self, key):
        """Hash function - division method"""
        if isinstance(key, int):
            return key % self.capacity
        # For strings
        hash_val = 0
        for char in str(key):
            hash_val = (hash_val * 31 + ord(char)) % self.capacity
        return hash_val
    
    def _resize(self):
        """Resize and rehash - O(n)"""
        old_buckets = self.buckets
        self.capacity *= 2
        self.buckets = [None] * self.capacity
        self.size = 0
        
        for bucket in old_buckets:
            current = bucket
            while current:
                self.put(current.key, current.value)
                current = current.next
    
    def put(self, key, value):
        """Insert/Update - O(1) average, O(n) worst"""
        index = self._hash(key)
        
        # Check if key exists
        current = self.buckets[index]
        while current:
            if current.key == key:
                current.value = value
                return
            current = current.next
        
        # Insert new node
        new_node = HashNode(key, value)
        new_node.next = self.buckets[index]
        self.buckets[index] = new_node
        self.size += 1
        
        # Check load factor
        if self.size > self.capacity * self.load_factor:
            self._resize()
    
    def get(self, key):
        """Get value - O(1) average, O(n) worst"""
        index = self._hash(key)
        current = self.buckets[index]
        
        while current:
            if current.key == key:
                return current.value
            current = current.next
        
        return None
    
    def remove(self, key):
        """Remove key - O(1) average, O(n) worst"""
        index = self._hash(key)
        current = self.buckets[index]
        prev = None
        
        while current:
            if current.key == key:
                if prev:
                    prev.next = current.next
                else:
                    self.buckets[index] = current.next
                self.size -= 1
                return True
            prev = current
            current = current.next
        
        return False
    
    def contains(self, key):
        """Check if key exists - O(1) average"""
        return self.get(key) is not None
    
    def __len__(self):
        return self.size


class HashTableOpenAddressing:
    """Hash Table with Linear Probing"""
    
    def __init__(self, capacity=10):
        self.capacity = capacity
        self.size = 0
        self.keys = [None] * capacity
        self.values = [None] * capacity
        self.DELETED = object()  # Marker for deleted entries
    
    def _hash(self, key):
        """Hash function"""
        if isinstance(key, int):
            return key % self.capacity
        hash_val = 0
        for char in str(key):
            hash_val = (hash_val * 31 + ord(char)) % self.capacity
        return hash_val
    
    def _probe(self, key):
        """Linear probing"""
        index = self._hash(key)
        initial_index = index
        
        while (self.keys[index] is not None and 
               self.keys[index] != self.DELETED and
               self.keys[index] != key):
            index = (index + 1) % self.capacity
            if index == initial_index:  # Table is full
                return -1
        return index
    
    def put(self, key, value):
        """Insert/Update - O(1) average"""
        if self.size >= self.capacity:
            return False
        
        index = self._probe(key)
        if index == -1:
            return False
        
        if self.keys[index] is None or self.keys[index] == self.DELETED:
            self.size += 1
        
        self.keys[index] = key
        self.values[index] = value
        return True
    
    def get(self, key):
        """Get value - O(1) average"""
        index = self._hash(key)
        initial_index = index
        
        while self.keys[index] is not None:
            if self.keys[index] == key:
                return self.values[index]
            index = (index + 1) % self.capacity
            if index == initial_index:
                break
        
        return None
    
    def remove(self, key):
        """Remove key - O(1) average"""
        index = self._hash(key)
        initial_index = index
        
        while self.keys[index] is not None:
            if self.keys[index] == key:
                self.keys[index] = self.DELETED
                self.values[index] = None
                self.size -= 1
                return True
            index = (index + 1) % self.capacity
            if index == initial_index:
                break
        
        return False


# Example usage
if __name__ == "__main__":
    # Chaining
    print("--- Hash Table with Chaining ---")
    ht = HashTable()
    ht.put("apple", 1)
    ht.put("banana", 2)
    ht.put("cherry", 3)
    
    print("Get 'apple':", ht.get("apple"))
    print("Contains 'banana':", ht.contains("banana"))
    print("Size:", len(ht))
    
    ht.remove("banana")
    print("After removing 'banana', contains:", ht.contains("banana"))
    
    # Open Addressing
    print("\n--- Hash Table with Open Addressing ---")
    ht2 = HashTableOpenAddressing()
    ht2.put(1, "one")
    ht2.put(2, "two")
    ht2.put(11, "eleven")  # Will probe
    
    print("Get 1:", ht2.get(1))
    print("Get 11:", ht2.get(11))
    ht2.remove(1)
    print("After removing 1, get 1:", ht2.get(1))

